debian/manpage/postfixadmin-cli.1
.TH "postfixadmin\-cli" "1" "2024-11-30" "PostfixAdmin" "postfixadmin-cli"
.SH Name
postfixadmin-cli \- a CLI tool to manage postfixadmin
.SH SYNOPSIS
\fBpostfixadmin-cli\fP \fI<module> <task>\fP \fB[options]\fP
.P
\fBpostfixadmin-cli\fP \fI<module>\fP \fBhelp\fP
.P
\fBpostfixadmin-cli\fP \fI<module> <task>\fP \fBhelp\fP
.SH DESCRIPTION
.B postfixadmin-cli
is a CLI tool to use postfixadmin like how you use the WebUI. It is a
shell sciprt wrapped around
\fI/usr/share/postfixadmin/scripts/postfixadmin-cli.php\fP.
.SH MODULES
.IP \fBadmin\fP
.IP \fBdomain\fP
.IP \fBmailbox\fP
.IP \fBalias\fP
.IP \fBaliasdomain\fP
.IP \fBfetchmail\fP
.SH TASKS
Most modules support the following tasks:
.IP \fBview\fP
View an item.
.IP \fBadd\fP
Add an item.
.IP \fBupdate\fP
Update an item.
.IP \fBdelete\fP
Delete an item.
.IP \fBscheme\fP
Print database scheme (useful for developers only).
.IP \fBhelp\fP
Print help output.
.SH OPTIONS
Different combinations of \fBMODULES\fP and \fBTASKS\fP have different
sets of available options to use. Please see the help messages and
the examples below.
.SH EXAMPLES
To add a mailbox interactively:
.P
.B postfixadmin-cli \fImailbox add\fP
.P
To add a mailbox non-interactively, with the new mailbox being inactive:
.P
.B postfixadmin-cli \fImailbox add\fP \fRnobody@example.com\fP
.B --password \fRdeadbeef12\fP --password2 \fRdeadbeef12\fP --name
.B \fR"Nobody Nogroup"\fP --welcome-mail --active \fR0\fP
.B --email-other \fRbackup@example.com\fP
.P
To view the information of a domain:
.P
.B postfixadmin-cli \fIdomain view\fP \fRexample.com\fP
.P
.SH REPORTING BUGS
Please report bugs to the Debian BTS, i.e. using \fBreportbug\fP(1).
.SH SEE ALSO
\fBpostfix\fP(1),
\fBpostconf\fP(1),
\fBpostmap\fP(1),
\fBmariadb\fP(1),
\fBpsql\fP(1),
\fBphp\fP(1)
.SH AUTHOR
Copyright 2024 Jing Luo.
.P
This manpage was written by Jing Luo, licensed under the terms of GNU
General Pulic License version 2, or at your option any later version.
There is absolutely no warranty; see \fI/usr/share/common-licenses/GPL-2\fR
for the terms of the license.


CHANGELOG.TXT
# Postfix Admin
#
# LICENSE
# This source file is subject to the GPL license that is bundled with
# this package in the file LICENSE.TXT.
#
# Further details on the project are available at https://github.com/postfixadmin/postfixadmin

MASTER (should have all changes from 3.3.10 plus)
------
   - Add Content Security Policy in common.php ( see https://github.com/postfixadmin/postfixadmin/issues/905, thanks @gitblacker ) 
   - Add SameSite, HttpOnly and (perhaps) secure parameters to the PHP session cookie ( see https://github.com/postfixadmin/postfixadmin/issues/903 - thanks @gitblacker)
   - Update moment.js library - see https://github.com/postfixadmin/postfixadmin/issues/902
   - Remove use of deprecated 'imap' module (see https://github.com/postfixadmin/postfixadmin/issues/812 and https://github.com/postfixadmin/postfixadmin/issues/812). This also removes the config for create_mailbox_subdirs etc.
   - Ability to limit alias destinations to only be for local domains (see https://github.com/postfixadmin/postfixadmin/pull/338/; thanks @michaelkrieger)
   - Ability to control control smtp access for a mailbox (so limiting outbound email, while still allowing incoming; see https://github.com/postfixadmin/postfixadmin/pull/796; thanks @johnfawcett)
   - MFA/TOTP support (see https://github.com/postfixadmin/postfixadmin/pull/753 - thank you @svenseeberg and @drunken-sod)
   - Update composer's PHP requirement to PHP 7.4+ (which matches Wordpress etc, PHP project is supporting 8.1<->8.3)
   - Make fetchmail.pl respect src_port (thanks @tempixx - see #784)
   - Improve github action (php 8.2)
   - Add Javascript username validation to prevent illegal characters (see #765; thanks @verdinago)
   - Fix typos / broken links etc (see #761; thanks @knofte)
   - Add initial support for OpenDKIM domain key handling (see #631; thanks @Freddo3000 )
   - vacation.pl updates - (log4perl warnings, typos etc, logging to file, localhost connection stuff - see https://github.com/postfixadmin/postfixadmin/discussions/648 thanks @puck)
   - fix display of active buttons in the bootstrap theme (see #749 - thanks @reinks2112)
   - fix theme logo + config interaction (see #718; thanks @dkatheininge)
   - fix pacrypt + php_crypt:MD5 
   - fix dovecot:CRYPT-METHOD (see #708; thanks @Coeur2Boeuf)
   - fix CSV alias export (see #703; thanks @dennis2society)
   - upgrade bundled jquery (v1.12.4 to v3.7.0) (fixes jquery ajax security problem, postfixadmin isn't vulnerable, see #734; thanks @tibor-banfalvi)
   - Require 'composer' for library installation (see install.sh)
   - Remove the dependency on 'doveadm pw' and support more hash mechanisms (see: #491)
   - Update debian/ patch (see #529)
   - Support $CONF['database_port'] when connecting to MySQL (see #549)
   - Add optional Dovecot mail-crypt plugin support - see https://github.com/postfixadmin/postfixadmin/issues/408
   - Add $CONF['site_url'] to allow administrators to override a detected site url (e.g. used in password recovery emails; see https://github.com/postfixadmin/postfixadmin/issues/446 )
   - Code reformat as PHPCS has a mind of it's own (function/method opening brace change of position)
   - Improved UTF8 support in vacation (see https://github.com/postfixadmin/postfixadmin/pull/484)
   - Fix quota levels losing config control (see bfc7af5c8efe2a68c47286cc870b56cb4f929a3f)
   - vacation.pl: improve autoreply detection (see https://github.com/postfixadmin/postfixadmin/pull/482 
   - vacation.pl: improve headers in auto-reply mails (add: "Auto-Submitted: auto-replied") see https://github.com/postfixadmin/postfixadmin/pull/483
   - vacation.pl: allow smtp helo to be customised; see https://github.com/postfixadmin/postfixadmin/pull/495
   - dark theme - use `$CONF['theme_css'] = 'css/dark-theme.css';` to enable - see https://github.com/postfixadmin/postfixadmin/issues/569, thanks @polymeer
   - support 'send email' using smtp+ssl, smtp+starttls or smtp - see https://github.com/postfixadmin/postfixadmin/pull/566, thanks @davidebeatrici
   - support smtpd_sender_login_maps via documents - see https://github.com/postfixadmin/postfixadmin/pull/565, thanks @davidebeatrici
   - fix database collation issues (try and use latin1_general_ci everywhere)

Version 3.3.13 - 2022/12/08
-------------------------------------------------
   - MySQL collation fix for quota/quota2 tables ( see https://github.com/postfixadmin/postfixadmin/issues/690 )
   - Fix MySQL not liking "alter table drop constraint ..." which seems to be MariaDB only (see #561 and #687) (thank you @stefanomarty)

Version 3.3.12 - 2022/12/04
-------------------------------------------------
   - MySQL collation setting change, see https://github.com/postfixadmin/postfixadmin/issues/595 (and #327 and #552)
   - Upgrade Smarty to 4.3.0 (see also #541)
   - Improve PHP 8.1/8.2 compatability (see #632)

Version 3.3.11 - 2022/03/02
-------------------------------------------------
   - Fix PHP 8 compatability for crypt() usage (see https://github.com/postfixadmin/postfixadmin/issues/547)
   - Support $CONF['database_port'] for MySQL databases (see https://github.com/postfixadmin/postfixadmin/issues/549 and https://github.com/postfixadmin/postfixadmin/issues/553)

Version 3.3.10 - 2021/08/09
-------------------------------------------------
   - Merge password expiration fixes from https://github.com/postfixadmin/postfixadmin/pull/493
   - Remove html readonly attribute from user's vacation page to/from selectors.
   - vacation.pl - allow smtp helo to be specified (see https://github.com/postfixadmin/postfixadmin/pull/495)
   - Security fix - ClickJacking protection (thanks @huntr-helper / @ranjit-git) (see https://github.com/postfixadmin/postfixadmin/issues/523)
   - Security fix (low risk) - Improve randomness with PFA_token for CSRF protection (thanks @michaellrowley)
   - Fix viewlog to allow admins to see all domains (thanks @pgimalac, https://github.com/postfixadmin/postfixadmin/issues/516)

scripts/postfixadmin-cli.php
#!/usr/bin/php
<?php

/**
 * Command-line code generation utility to automate administrator tasks.
 *
 * Shell dispatcher class
 *
 * PHP versions 4 and 5
 *
 * CakePHP(tm) :  Rapid Development Framework <http://www.cakephp.org/>
 * Copyright 2005-2008, Cake Software Foundation, Inc.
 *                                                              1785 E. Sahara Avenue, Suite 490-204
 *                                                              Las Vegas, Nevada 89104
 * Modified for PostfixAdmin by Valkum 2011
 * Modified for PostfixAdmin by Christian Boltz 2011-2013
 *
 * Copyright 2010
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @filesource
 * @copyright           Copyright 2005-2008, Cake Software Foundation, Inc.
 * @link                                http://postfixadmin.sourceforge.net/ Postfixadmin on Sourceforge
 * @package                     postfixadmin
 * @subpackage          -
 * @since                       -
 * @version                     $Revision$
 * @modifiedby          $LastChangedBy$
 * @lastmodified        $Date$
 * @license                     http://www.opensource.org/licenses/mit-license.php The MIT License
 */


class PostfixAdmin
{
    /**
     * Version
     *
     * @var string
     */
    public $version = '0.3';

    /**
     * Standard input stream.
     *
     * @var resource
     */
    public $stdin;

    /**
     * Standard output stream.
     *
     * @var resource
     */
    public $stdout;

    /**
     * Standard error stream.
     *
     * @var resource
     */
    public $stderr;

    /**
     * Contains command switches parsed from the command line.
     *
     * @var array
     */
    public $params = array();

    /**
     * Contains arguments parsed from the command line.
     *
     * @var array
     */
    public $args = array();

    /**
     * The file name of the shell that was invoked.
     *
     * @var string
     */
    public $shell;

    /**
     * The class name of the shell that was invoked.
     *
     * @var string
     */
    public $shellClass;

    /**
     * The command called if public methods are available.
     *
     * @var string
     */
    public $shellCommand;

    /**
     * The name of the shell in camelized.
     *
     * @var string
     */
    public $shellName;

    /**
     * Constructor
     *
     * @param array $args the argv.
     */
    public function __construct($args = array())
    {
        set_time_limit(0);
        $this->__initConstants();
        $this->parseParams($args);
        $this->__initEnvironment();
    }

    /**
     * Defines core configuration.
     */
    private function __initConstants()
    {
        ini_set('display_errors', '1');
        ini_set('error_reporting', '' . E_ALL);
        ini_set('html_errors', "0");
        ini_set('implicit_flush', "1");
        ini_set('max_execution_time', "0");
    }

    /**
     * Defines current working environment.
     */
    private function __initEnvironment()
    {
        $this->stdin  = fopen('php://stdin', 'r');
        $this->stdout = fopen('php://stdout', 'w');
        $this->stderr = fopen('php://stderr', 'w');

        if (basename(__FILE__) !=  basename($this->args[0])) {
            $this->stderr('Warning: the dispatcher may have been loaded incorrectly, which could lead to unexpected results...');
            if ($this->getInput('Continue anyway?', array('y', 'n'), 'y') == 'n') {
                exit(1);
            }
        }

        $this->shiftArgs();
    }

    /**
     * postfixadmin-cli admin view admin@example.com
     *              - Create AdminHandler.
     *              - and then a CliView object (Shell class)
     *              - call CliView->view() ... which under the covers uses AdminHandler*
     */
    public function dispatch()
    {
        check_db_version(); # ensure the database layout is up to date

        if (!isset($this->args[0])) {
            $this->help();
            return 1;
        }

        $this->shell = $this->args[0];
        $this->shiftArgs();
        $this->shellName = ucfirst($this->shell);
        $this->shellClass = $this->shellName . 'Handler';


        if ($this->shell == 'help') {
            $this->help();
            return 1;
        }

        $command = $this->args[0];

        $this->shellCommand = $command;
        $this->shellClass = 'Cli' . ucfirst($command);

        if (ucfirst($command) == 'Add' || ucfirst($command) == 'Update') {
            $this->shellClass = 'CliEdit';
        }

        if (!class_exists($this->shellClass)) {
            $this->stderr('Unknown task ' . $this->shellCommand);
            return 1;
        }

        $shell = new $this->shellClass($this);

        $shell->handler_to_use = ucfirst($this->shell) . 'Handler';

        if (!class_exists($shell->handler_to_use)) {
            $this->stderr('Unknown module ' . $this->shell);
            return 1;
        }

        $task = ucfirst($command);

        $shell->new = 0;
        if ($task == 'Add') {
            $shell->new = 1;
        }

        # TODO: add a way to Cli* to signal if the selected handler is supported (for example, not all *Handler support changing the password)

        if (strtolower(get_parent_class($shell)) == 'shell') {
            $handler = new $shell->handler_to_use();
            if (in_array($task, $handler->taskNames)) {
                $this->shiftArgs();
                $shell->startup();

                if (isset($this->args[0]) && $this->args[0] == 'help') {
                    if (method_exists($shell, 'help')) {
                        $shell->help();
                        return 1;
                    } else {
                        $this->help();
                        return 1;
                    }
                }

                return $shell->execute();
            }
        }

        $classMethods = get_class_methods($shell);

        $privateMethod = $missingCommand = false;
        if ((in_array($command, $classMethods) || in_array(strtolower($command), $classMethods)) && strpos($command, '_', 0) === 0) {
            $privateMethod = true;
        }

        if (!in_array($command, $classMethods) && !in_array(strtolower($command), $classMethods)) {
            $missingCommand = true;
        }

        $protectedCommands = array(
            'in', 'out', 'err', 'hr', 'log',
            '__construct', 'dispatch', 'stdout', 'stderr'
        );

        if (in_array(strtolower($command), $protectedCommands)) {
            $missingCommand = true;
        }

        if ($missingCommand && method_exists($shell, 'main')) {
            $shell->startup();
            return $shell->main();
        } elseif (!$privateMethod && method_exists($shell, $command)) {
            $this->shiftArgs();
            $shell->startup();
            return $shell->{$command}();
        } else {
            $this->stderr("Unknown {$this->shellName} command '$command'.\nFor usage, try 'postfixadmin-cli {$this->shell} help'.\n\n");
            return 1;
        }
    }

    /**
     * Prompts the user for input, and returns it.
     *
     * @param string $prompt Prompt text.
     * @param mixed $options Array or string of options.
     * @param string $default Default input value.
     * @return string Either the default value, or the user-provided input.
     */
    public function getInput($prompt, $options = null, $default = null)
    {
        if (!is_array($options)) {
            $print_options = '';
        } else {
            $print_options = '(' . implode('/', $options) . ')';
        }

        if ($default == null) {
            $this->stdout($prompt . " $print_options \n" . '> ', false);
        } else {
            $this->stdout($prompt . " $print_options \n" . "[$default] > ", false);
        }
        $result = fgets($this->stdin);

        if ($result === false) {
            exit(1);
        }
        $result = trim($result);

        if ($default != null && empty($result)) {
            return $default;
        }
        return $result;
    }

    /**
     * Outputs to the stdout filehandle.
     *
     * @param string $string String to output.
     * @param boolean $newline If true, the outputs gets an added newline.
     */
    public function stdout($string, $newline = true)
    {
        if ($newline) {
            fwrite($this->stdout, $string . "\n");
        } else {
            fwrite($this->stdout, $string);
        }
    }

    /**
     * Outputs to the stderr filehandle.
     *
     * @param string $string Error text to output.
     */
    public function stderr($string)
    {
        fwrite($this->stderr, 'Error: '. $string . "\n");
    }

    /**
     * Parses command line options
     *
     * @param array $params Parameters to parse
     */
    public function parseParams($params)
    {
        $count = count($params);
        for ($i = 0; $i < $count; $i++) {
            if ($params[$i] != '' && $params[$i][0] === '-' && $params[$i] != '-1') {
                $key = substr($params[$i], 1);
                if (isset($params[$i + 1])) {
                    # TODO: ideally we should know if a parameter can / must have a value instead of whitelisting known valid values starting with '-' (probably only bool doesn't need a value)
                    if ($params[$i + 1][0] === '-' && $params[$i + 1] != '-1') {
                        $this->params[$key] = true;
                    } else {
                        $this->params[$key] = $params[$i + 1];
                        $i++;
                    }
                }
            } else {
                $this->args[] = $params[$i];
            }
        }
    }

    /**
     * Removes first argument and shifts other arguments up
     *
     * @return boolean False if there are no arguments
     */
    public function shiftArgs()
    {
        if (empty($this->args)) {
            return false;
        }
        unset($this->args[0]);
        $this->args = array_values($this->args);
        return true;
    }

    /**
     * prints help message and exits.
     */
    public function help()
    {
        $this->stdout("\nWelcome to Postfixadmin-CLI v" . $this->version);
        $this->stdout("---------------------------------------------------------------");
        $this->stdout("Usage:");
        $this->stdout("    postfixadmin-cli <module> <task> [--option value --option2 value]");
        $this->stdout("");
        $this->stdout("Available modules:");

        $modules = explode(',', 'admin,domain,mailbox,alias,aliasdomain,fetchmail');
        foreach ($modules as $module) {
            $this->stdout("    $module");
        }
        $this->stdout("");
        $this->stdout("Most modules support the following tasks:");
        $this->stdout("    view      View an item");
        $this->stdout("    add       Add an item");
        $this->stdout("    update    Update an item");
        $this->stdout("    delete    Delete an item");
        $this->stdout("    scheme    Print database scheme (useful for developers only)");
        $this->stdout("    help      Print help output");
        $this->stdout("");
        $this->stdout("");
        $this->stdout("For module-specific help, see:");
        $this->stdout("");
        $this->stdout("    postfixadmin-cli <module> help");
        $this->stdout("        print a detailed list of available commands");
        $this->stdout("");
        $this->stdout("    postfixadmin-cli <module> <task> help");
        $this->stdout("        print a list of available options.");
        $this->stdout("");

        exit();
    }
}


define("POSTFIXADMIN_CLI", 1);

require_once(dirname(__FILE__) . '/../common.php');

$dispatcher = new PostfixAdmin($argv);
try {
    $retval = $dispatcher->dispatch();
} catch (Exception $e) {
    $dispatcher->stderr("Execution Exception: " . $e->getMessage());
    $retval = 1;
}
exit($retval);

/* vim: set expandtab softtabstop=4 tabstop=4 shiftwidth=4: */
